src/arch/metal.rs

```diff
@@ -780,6 +788,19 @@ zero_bss_loop:
     j zero_bss_loop
 zero_bss_done:
 
+    // Say to other harts that the initialization is done.
+    // This is atomic and memory is ordered in a way that the 
+    // initialization will then be visible as soon as the
+    // boolean is reset.
+    amoswap.w.rl x0, x0, (t3)   
+    j end_wait
+    
+    // Wait until initialization is done
+wait_bss_end:
+    lw t1, (t3)
```

Comment:
Here you need an atomic too, probably with acquire semantic (you should always pair a release with an acquire to guarantee ordering, at least on other architecture I know about).

Reply:
Actually I realize RISC-V might not call those atomics, but maybe just loads with acquire semantic?

Reply:
I can replace `amoswap.w.rl` by `amoswap.w.aqrl`. This means that the operation will be done in a sequential order in respect to the BSS zeroing. It will be visible by other hart in a sequential order too.

We do not need the other harts to acquire a lock, but only to pull the boolean until it is set to 0 by the boot hart.
The boolean is set to 1 by default, and cleared only by hart 0. We need to ensure that the BSS zeroing is done and visible before the boolean clear. So sequential ordering of the boolean clear should work, right?

Spec:

> To help implement multiprocessor synchronization, the AMOs optionally provide release consis-
tency semantics. If the aq bit is set, then no later memory operations in this RISC-V hart can be
observed to take place before the AMO. Conversely, if the rl bit is set, then other RISC-V harts will
not observe the AMO before memory accesses preceding the AMO in this RISC-V hart. Setting
both the aq and the rl bit on an AMO makes the sequence sequentially consistent, meaning that it
cannot be reordered with earlier or later memory operations from the same hart.

Reply:
The acquire and release semantic are local to a hart, by themselves they do not change what other harts can see.

Here the release semantic means that every operations before the AMO will be ordered before the AMO, while the acquire means that no operation that happens after the AMO will be ordered before the AMO. But you have to think of that as the other in which the hart will submit the changes to the memory system (caches, etc...).

To see a consistent ordering between two harts we need two things:
1. The hart doing the writes needs to decide on its own ordering (the order in which it submits changes to the memory subsystem).
2. The hart doing the reads needs to check if there were updated on the value it reads, so it needs to ask the memory subsystem if anything changed and in which order.

You are doing (1), but not (2), because you don't use a fence or atomic operation to read the value from memory.
Feel free to ask if anything is unclear :)